commit

DzmitryMiadzvedzeu opened the pull request.

DzmitryMiadzvedzeu committed:
commit

apsent92 commented on pom.xml:
```diff
@@ -0,0 +1,153 @@
+            <artifactId>postgresql</artifactId>
+            <version>42.2.20</version>
+        </dependency>
+<!--
```
уберешь закоменченый код

apsent92 commented on src/main/java/org/doc/com/api_errors/ApiError.java:
```diff
@@ -0,0 +1,13 @@
+@Getter
+@Setter
+@AllArgsConstructor
+public class ApiError {
```
наследоватся от RuntimeException нужно

apsent92 commented on src/main/java/org/doc/com/controller/DocumentController.java:
```diff
@@ -0,0 +1,93 @@
+import java.util.stream.Collectors;
+
+
+//http://localhost:8080/v1/documents (for postman)
```
закоменченый код убрать

apsent92 commented on src/main/java/org/doc/com/controller/DocumentController.java:
```diff
@@ -0,0 +1,93 @@
+    @PostMapping
+    public ResponseEntity<DocumentDto> create(@Valid @RequestBody DocumentCreateDtoWithToken documentCreateDtoWithToken) {
+        if (documentCreateDtoWithToken.getTransactionToken() == null || documentCreateDtoWithToken.getTransactionToken().isEmpty()) {
+            throw new IllegalArgumentException("Transaction token is required");
```
обработка в секюрити конфиге должна быть, и общие методы прокида ошибок оп секьюрити

apsent92 replied:
мб  стандартную аутентификацию спринга может лучше заюзай, передавать параметром токе в сервис хреновая практика - нарушается сингл респонсибилитис

DzmitryMiadzvedzeu replied:
это не секьюрити, я вообще аутентификацию не делал. Ты мне когда-то сказал про идемпотентность, это как раз тот метод который который не идемпотентен из коробки.

apsent92 commented on src/main/java/org/doc/com/controller/DocumentController.java:
```diff
@@ -0,0 +1,93 @@
+            throw new IllegalArgumentException("Transaction token is required");
+        }
+        Document document = mapper.documentCreateWithToken(documentCreateDtoWithToken);
+        Document createdDocument = service.create(document, documentCreateDtoWithToken.getTransactionToken());
```
предлагаю переменную назвать result, ниже тоже

apsent92 commented on src/main/java/org/doc/com/controller/DocumentController.java:
```diff
@@ -0,0 +1,93 @@
+            @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)LocalDateTime createdAt,
+            @RequestParam(required = false)
+            @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)LocalDateTime updatedAt) {
+        List<Document> documents = service.search(serialNumber, title, status, type,
```
параметры метода собери в DocumentFilter и через  билдер (аннотация @Builder от ломбока) собири в этом методе

apsent92 commented on src/main/java/org/doc/com/controller/DocumentController.java:
```diff
@@ -0,0 +1,93 @@
+        List<Document> documents = service.search(serialNumber, title, status, type,
+                createdAt, updatedAt);
+        List<DocumentDto> documentDtos = documents.stream()
+                .map(mapper :: toDto).collect(Collectors.toList());
```
стримы пиши по правилу строчка - точка : читабельность увеличится

apsent92 commented on src/main/java/org/doc/com/controller/DocumentController.java:
```diff
@@ -0,0 +1,93 @@
+
+    @GetMapping
+    public ResponseEntity<List<DocumentDto>> getAll() {
+        List<DocumentDto> documentDtos = service.getAll().stream()
```
строчка - точка

apsent92 commented on src/main/java/org/doc/com/controller/DocumentController.java:
```diff
@@ -0,0 +1,93 @@
+
+    @GetMapping("/{id}")
+    public ResponseEntity<DocumentDto> getById(@PathVariable Long id) {
+        return ResponseEntity.ok(mapper.toDto(service.findById(id)));
```
service.findById(id) вынеси в переменную - вложенность трудно читается

apsent92 commented on src/main/java/org/doc/com/dto/DocumentCreateDto.java:
```diff
@@ -0,0 +1,22 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class DocumentCreateDto {
```
DocumentCreateRequest - и мапперы с переменными переименуй

apsent92 commented on src/main/java/org/doc/com/controller/DocumentController.java:
```diff
@@ -0,0 +1,93 @@
+
+    @PutMapping("/{id}")
+    public ResponseEntity<DocumentDto> update(@PathVariable Long id,
+                                              @Valid @RequestBody DocumentCreateDto documentCreateDto) {
```
пусть все параметры в ендпойнтах будут наз-ся request

apsent92 commented on src/main/java/org/doc/com/dto/DocumentCreateDto.java:
```diff
@@ -0,0 +1,22 @@
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
```
зачем конструкторы?

apsent92 commented on src/main/java/org/doc/com/dto/DocumentCreateDto.java:
```diff
@@ -0,0 +1,22 @@
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
зачем это?

apsent92 commented on src/main/java/org/doc/com/dto/DocumentCreateDtoWithToken.java:
```diff
@@ -0,0 +1,24 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class DocumentCreateDtoWithToken {
```
логично было бы наследоваться от класса выше, но опять же, передавать токен в реквесте и потом в сервис плохая практика

apsent92 commented on src/main/java/org/doc/com/entity/Document.java:
```diff
@@ -0,0 +1,55 @@
+
+@Entity
+@Table(name = "Documents")
+@Data
```
лишняя анноташка вроде

apsent92 commented on src/main/java/org/doc/com/entity/Document.java:
```diff
@@ -0,0 +1,55 @@
+@Entity
+@Table(name = "Documents")
+@Data
+@NoArgsConstructor
```
тоже вроде есть в @Entity, почитай что она содержит и убери лишнее

apsent92 commented on src/main/java/org/doc/com/entity/Document.java:
```diff
@@ -0,0 +1,55 @@
+    private Long id;
+
+    @NotBlank(message = "Serial number please")
+    @Pattern(regexp = "^[A-Z]{2}\\d{5}-((EC)|(HR)|(DE)|(MA))$",
```
круто что заморочился) ненавижу regex

apsent92 commented on src/main/java/org/doc/com/entity/Document.java:
```diff
@@ -0,0 +1,55 @@
+    private String serialNumber;
+
+    @NotBlank(message = "Title can not be empty")
+    @Length(max = 40, message = "This field mustn't be longer than 40 characters")
```
для чего? вопрос не убрать это, а ответить, я хз что за функционал

DzmitryMiadzvedzeu replied:
Для того, чтобы в таблице небыло километрового заголовка

apsent92 commented on src/main/java/org/doc/com/enums/Status.java:
```diff
@@ -0,0 +1,7 @@
+package org.doc.com.enums;
+
+public enum Status {
```
смысл этого статуса?

DzmitryMiadzvedzeu replied:
когда происходит изменение документа, у него в меняется статус с CREATED на UPDATED также присваивается время апдэйта (дата создания остаётся). А потом можно даже сделать выборку по статусу и получить список ранжированный по дате обновления

apsent92 replied:
из какой либо бизнес логики не виже смысла применения, не важно отредактированый документ или нет

для статусов я бы предложил сделать: черновик, на рассмотрении, утвержден, отклонен  - я думаю это обязательно нужно сделать, потому что система документооборота это подразумевает
Сделать в контроллере методы перехода в каджый статус (кроме черновик - это будет создание) и проверки по схеме:
чернови -> на рассмотрении -> утвержден ИЛИ отклонен
отдельно из отклонен переход в статус черновик через метод апдейт

редактировать можно документы только в статусе черновик

apsent92 commented on src/main/java/org/doc/com/service/DocumentServiceImpl.java:
```diff
@@ -0,0 +1,80 @@
+    private final TransactionTokenJpaRepository tokenRepository;
+
+    @Override
+    public Document create(Document document, String transactionToken) {
```
транзакция

apsent92 commented on src/main/java/org/doc/com/service/DocumentServiceImpl.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Transactional
+    @Override
+    public Document update(Long id, Document document) {
```
транзакция

apsent92 commented on src/main/java/org/doc/com/service/DocumentServiceImpl.java:
```diff
@@ -0,0 +1,80 @@
+            throw new DocumentExistsException("Document with this number already exists");
+        }
+        repository.save(document);
+        tokenRepository.save(new TransactionToken(transactionToken, LocalDateTime.now()));
```
и смысл вообще сохранять токен

apsent92 commented on src/main/java/org/doc/com/controller/DocumentController.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @PutMapping("/{id}")
+    public ResponseEntity<DocumentDto> update(@PathVariable Long id,
```
для апдейта создаай отдельный класс наследник от DocumentCreateDto с полем id и над ним аннотацию @NotBlank (от пакета validation, не от ломбока)

apsent92 commented on src/main/java/org/doc/com/dto/DocumentCreateDto.java:
```diff
@@ -0,0 +1,22 @@
+
+    private String title;
+
+    private Type type;
```
я так понимаю должно быть обязательное? если да, то анноташку @NotNull

apsent92 commented on src/main/java/org/doc/com/dto/DocumentCreateDto.java:
```diff
@@ -0,0 +1,22 @@
+
+    private String serialNumber;
+
+    private String title;
```
я так понимаю тоже обязательное для заполнения, анноташку @NotBlank

apsent92 commented on src/main/java/org/doc/com/entity/Document.java:
```diff
@@ -0,0 +1,55 @@
+    @UpdateTimestamp
+    private LocalDateTime updatedAt;
+
+    public Document(String serialNumber, String title, Type type) {
```
зачем конструктор с такимнабором параметров?

apsent92 commented on src/main/java/org/doc/com/entity/Document.java:
```diff
@@ -0,0 +1,55 @@
+        this.createdAt = LocalDateTime.now();
+        this.updatedAt = LocalDateTime.now();
+    }
+}
```
@PrePersist
	public void prePersist()
	{
		created = LocalDateTime.now();
		updated = LocalDateTime.now();
	}

	@PreUpdate
	public void preUpdate()
	{
		updated = LocalDateTime.now();
	}
	используй такое для сетания created и updated - это будет работать при сохранении в бд и тебе не нужно больше думать об этих полях

apsent92 commented on src/main/java/org/doc/com/entity/Document.java:
```diff
@@ -0,0 +1,55 @@
+    @Temporal(TemporalType.TIMESTAMP)
+    private LocalDateTime createdAt = LocalDateTime.now();
+
+    @Temporal(TemporalType.TIMESTAMP)
```
это лишнее, проверь, должно без этого работать

apsent92 commented on src/main/java/org/doc/com/service/DocumentServiceImpl.java:
```diff
@@ -0,0 +1,80 @@
+        Document existingDocument = repository.findById(id)
+                .orElseThrow(() -> new DocumentNotFoundException("Document with this id: "
+                        + id + " not found"));
+        existingDocument.setSerialNumber(document.getSerialNumber());
```
это поле не должно изменятся

apsent92 commented on src/main/java/org/doc/com/entity/Document.java:
```diff
@@ -0,0 +1,55 @@
+@Table(name = "Documents")
+@Data
+@NoArgsConstructor
+public class Document {
```
очень здорово было бы сделать хранение вложений (ворд файлы или пдф) но это жирно будет для такого срока и без опыта, поэтому прделагаю поле text на 2000 символов сделать обязательное, которое будет хранить информацию документа

и прямо так в коде и напишешь //TODO вложение Attachment

DzmitryMiadzvedzeu committed:
Changes:
ApiError.java, application.properties, Department.java, Document.java, DocumentController.java, DocumentCreateDto.java, DocumentCreateDtoWithToken.java,       DocumentDto.java, DocumentExistsException.java, DocumentMapper.java, DocumentNotFoundException.java, DocumentService.java, DocumentServiceImpl.java, DocumentSpecifications.java, pom.xml,  Status.java, TransactionToken.java,           TransactionTokenJpaRepository.java,
Added: all for new entity, exceptions, changelogs, security.
CustomUserDetailsService.java
db.changelog-master.xml
document-changelock.xml
SecurityConfig.java
User.java
user-changelock.xml
UserController.java
UserCreateDto.java
UserDto.java
UserJpaRepository.java
UserMapper.java
UserNotFoundException.java
UserPrincipal.java
UserService.java
UserServiceImpl.java
